exporter/dynatraceexporter/internal/serialization/histogram.go

```diff
@@ -58,64 +53,107 @@ func serializeHistogram(name, prefix string, dims dimensions.NormalizedDimension
+	if dp.HasMin() && dp.HasMax() && dp.HasSum() {
+		return dp.Min(), dp.Max(), dp.Sum()
 	}
 
-	// Because we do not know the actual min and max, we estimate them based on the min and max non-empty bucket
-	minIdx, maxIdx := -1, -1
-	for y := 0; y < len(counts); y++ {
-		if counts[y] > 0 {
-			if minIdx == -1 {
-				minIdx = y
-			}
-			maxIdx = y
-		}
+	// a single-bucket histogram is a special case
+	if len(counts) == 1 {
+		return estimateSingleBucketHistogram(dp)
 	}
 
-	if minIdx == -1 || maxIdx == -1 {
-		return 0, 0
-	}
+	foundNonEmptyBucket := false
+	var min, max, sum float64 = 0, 0, 0
 
-	var min, max float64
+	// Because we do not know the actual min, max, or sum, we estimate them based on non-empty buckets
+	for i := 0; i < len(counts); i++ {
+		// empty bucket
+		if counts[i] == 0 {
+			continue
+		}
 
-	if dp.HasMin() {
-		min = dp.Min()
-	} else {
-		// Use lower bound for min unless it is the first bucket which has no lower bound, then use upper
-		if minIdx == 0 {
-			min = bounds[minIdx]
+		// range for counts[i] is bounds[i-1] to bounds[i]
+
+		// min estimation
+		if !foundNonEmptyBucket {
+			foundNonEmptyBucket = true
+			if i == 0 {
+				// if we're in the first bucket, the best estimate we can make for min is the upper bound
+				min = bounds[i]
+			} else {
+				min = bounds[i-1]
+			}
+		}
+
+		if i == len(counts)-1 {
+			// if we're in the last bucket, the best estimate we can make for max is the lower bound
+			max = bounds[i-1]
 		} else {
-			min = bounds[minIdx-1]
+			max = bounds[i]
+		}
+
+		switch i {
+		case 0:
+			// in the first bucket, estimate sum using the upper bound
+			sum += float64(counts[i]) * bounds[i]
+		case len(counts) - 1:
+			// in the last bucket, estimate sum using the lower bound
+			sum += float64(counts[i]) * bounds[i-1]
+		default:
+			// in any other bucket, estimate sum using the bucket midpoint
+			sum += float64(counts[i]) * (bounds[i] + bounds[i-1]) / 2
 		}
 	}
 
+	if dp.HasMin() {
+		min = dp.Min()
+	}
 	if dp.HasMax() {
 		max = dp.Max()
-	} else {
-		// Use upper bound for max unless it is the last bucket which has no upper bound, then use lower
-		if maxIdx == len(counts)-1 {
-			max = bounds[maxIdx-1]
-		} else {
-			max = bounds[maxIdx]
-		}
+	}
+	if dp.HasSum() {
+		sum = dp.Sum()
 	}
 
 	// Set min to average when higher than average. This can happen when most values are lower than first boundary (falling in first bucket).
 	// Set max to average when lower than average. This can happen when most values are higher than last boundary (falling in last bucket).
-	avg := dp.Sum() / float64(dp.Count())
+	avg := sum / float64(dp.Count())
 	if min > avg {
 		min = avg
 	}
 	if max < avg {
 		max = avg
 	}
 
-	return min, max
+	return min, max, sum
+}
+
+func estimateSingleBucketHistogram(dp pmetric.HistogramDataPoint) (float64, float64, float64) {
+	min, max, sum := 0.0, 0.0, 0.0
+
+	if dp.HasSum() {
+		sum = dp.Sum()
+	}
+
+	mean := sum / float64(dp.Count())
```

Comment:
do we have to check for Count == 0? I think above we check if there is exactly one bucket, but not if the value in the bucket is > 0, or am I missing something? Otherwise we might divide by 0.

Reply:
https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/11044/files#diff-ff59d8892cc874e319200c4bcdba980219543eec16aa5ccb3ef19170952a0456R35

The function is only called for histograms with non-zero counts. I can document this requirement